src/test/java/com/synopsys/integration/jenkins/detect/service/DetectArgumentServiceTest.java

```diff
@@ -0,0 +1,203 @@
+
+    private DetectArgumentService detectArgumentService;
+    private Map<String, String> inputDetectProperties;
+    private Map<String, String> expectedVisibleDetectProperties;
+    private Map<String, String> expectedHiddenDetectProperties;
+
+    @BeforeEach
+    public void setUp() {
+        // Setup logger
+        TaskListener taskListener = Mockito.mock(TaskListener.class);
+        JenkinsIntLogger jenkinsIntLogger = new JenkinsIntLogger(taskListener);
+
+        jenkinsIntLogger.setLogLevel(LogLevel.DEBUG);
+        intEnvironmentVariables.put("DETECT_PLUGIN_ESCAPING", "true");
+
+        // Set default expected values. Tests will apply different if needed
+        Mockito.when(taskListener.getLogger()).thenReturn(new PrintStream(byteArrayOutputStream));
+        Mockito.when(jenkinsVersionHelper.getJenkinsVersion()).thenReturn(Optional.of(expectedJenkinsVersion));
+        Mockito.when(jenkinsVersionHelper.getPluginVersion(pluginName)).thenReturn(Optional.of(expectedJenkinsPluginVersion));
+
+        // These are the standard input detect properties
+        inputDetectProperties = new LinkedHashMap<>();
+        inputDetectProperties.put("--detect.docker.passthrough.service.timeout", "1234567890");
+        inputDetectProperties.put("--detect.cleanup", "ThisIsFalse");
+
+        // These are the standard expected detect properties. Tests will update if needed.
+        // They SHOULD be visible in the output log message and be within the return List
+        expectedVisibleDetectProperties = new LinkedHashMap<>();
+        expectedVisibleDetectProperties.putAll(inputDetectProperties);
+        expectedVisibleDetectProperties.put(loggingLevelKey, jenkinsIntLogger.getLogLevel().toString());
+
+        // These are the automatically appended detect properties added by DetectArgumentService(). Tests will update if needed.
+        // They should NOT be visible in the output log message but should be within the return List
+        expectedHiddenDetectProperties = new LinkedHashMap<>();
+        expectedHiddenDetectProperties.put(jenkinsVersionParam, expectedJenkinsVersion);
+        expectedHiddenDetectProperties.put(pluginVersionParam, expectedJenkinsPluginVersion);
+
+        // Create object in test
+        detectArgumentService = new DetectArgumentService(jenkinsIntLogger, jenkinsVersionHelper);
+    }
+
+    private String createDetectPropertiesInputString() {
+        return inputDetectProperties.keySet().stream()
+                   .map(key -> key + "=" + inputDetectProperties.get(key))
+                   .collect(Collectors.joining(" "));
+    }
+
+    @Test
+    public void testGetDetectArguments() {
+        List<String> detectCommandLine = detectArgumentService.getDetectArguments(intEnvironmentVariables, strategyEscaper, invocationParameters, createDetectPropertiesInputString());
+
+        assertEquals(6, detectCommandLine.size()); // Invocation args (1) + passed args (2) + auto added args (3)
```

Comment:
You have comments in a lot of places for the size-- perhaps we should extract them into static variables? That way if we add an arg in the future we can just update a number in one place instead of updating all of the tests.

Something like:
EXPECTED_INVOCATION_ARG_COUNT + DEFAULT_ARG_COUNT + PHONE_HOME_ARG_COUNT + LOGGING_ARG_COUNT

Reply:
I think doing this adds unnecessary complexity. The comments were added to aid someone reviewing/supporting the tests, but the values can (and often are) different.

  | Invocation | Passed | Auto | TOTAL
-- | -- | -- | -- | --
testGetDetectArguments() | 1 | 2 | 3 | 6
testShouldEscapeFalse() | 1 | 2 | 3 | 6
testVariableReplaced() | 1 | 3 | 3 | 7
testVariableNotReplaced() | 1 | 3 | 3 | 7
testCustomLogLevel() | 1 | 3 | 2 | 6
testNoDetectArguments() | 1 | 0 | 3 | 4
testUnknownJenkinsVersion() | 1 | 2 | 3 | 6
testEmptyIntEnvironmentVariables() | 1 | 2 | 3 | 6

4 of the tests: testGetDetectArguments, testShouldEscapeFalse, testUnknownJenkinsVersion, testEmptyIntEnvironmentVariables are the same. The other 4 are different and for different reasons.

Reply:
I messed around with it a little bit and I'm inclined to agree with you-- adding variables for counts of arguments doesn't seem to ultimately provide the clarity I was hoping it would.